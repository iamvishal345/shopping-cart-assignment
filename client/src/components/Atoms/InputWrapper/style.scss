@mixin label-active {
  top: -5px;
  font-size: 11px;
  left: 0px;
  color: #62c9d8;
}
.form-input {
  position: relative;
  width: 98%;
  height: 1rem;
  padding: 20px 0;
  input {
    width: 100%;
    border: none;
    border-bottom: 1px solid #989290;
    &:focus {
      outline: none;
      border-bottom: 2px solid #62c9d8;
    }
  }
  .form-input-label {
    font-size: 12px;
    position: absolute;
    pointer-events: none;
    left: 0px;
    top: 15px;
    transition: 300ms ease all;
    color: #79716f;
    &.input-value {
      @include label-active();
    }
  }
  > input:focus {
    ~ .form-input-label {
      @include label-active();
    }
  }
}
