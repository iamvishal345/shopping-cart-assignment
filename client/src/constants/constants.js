export const SIGNIN = {
  header: "Login",
  subHeader: "Get access to your Orders, Wishlist and Recommendayions",
  fields: [
    {
      name: "email",
      type: "email",
      id: "email",
      label: "Email",
    },
    {
      name: "password",
      type: "password",
      id: "password",
      label: "Password",
    },
  ],
};

export const REGISTER = {
  header: "Signup",
  subHeader: "We do not share your personal details with anyone",
  fields: [
    {
      name: "firstName",
      type: "text",
      id: "firstName",
      label: "FirstName",
    },
    {
      name: "lastName",
      type: "text",
      id: "lastName",
      label: "LastName",
    },
    {
      name: "email",
      type: "email",
      id: "email",
      label: "Email",
    },
    {
      name: "password",
      type: "password",
      id: "password",
      label: "Password",
    },
    {
      name: "confirmPassword",
      type: "password",
      id: "confirmPassword",
      label: "Confirm Password",
    },
  ],
};

export const PRODUCT_CATEGORIES = [
  {
    header: "Fruits & Vegetables",
    details: "A variety of fresh fruits and vegetables.",
    link: "fruit-and-veg",
    image: "fruits.png",
    imageAlt: "A Basket of fresh fruits and vegetables",
  },
  {
    header: "Bakery Cakes and Dairy",
    details:
      "The best cupcakes, cookies, cakes, pies, cheesecakes, fresh bread, biscuits, muffins, bagels, fresh coffee, milk and more.",
    link: "bakery-cakes-dairy",
    image: "bakery.png",
    imageAlt: "Slices of White bread",
  },
  {
    header: "Beverages",
    details:
      "Our beverage department will ensure your fridge is always fully stocked. Shop for soda, juice, beer and more.",
    link: "beverages",
    image: "beverages.png",
    imageAlt: "Jar of Coffee, Packet of Tea and Carton of Juice ",
  },
  {
    header: "Beauty and Hygiene",
    details:
      "Buy beauty and personal care products online in India at best price.",
    link: "beauty-hygiene",
    image: "beauty.png",
    imageAlt: "Small bottles of cosmetics and Skin care",
  },
  {
    header: "Baby Care",
    details: "Shop online for Baby Products, Diapers, Skin Care Products etc.",
    link: "baby-care",
    image: "baby.png",
    imageAlt: "Baby in bathtub",
  },
];
