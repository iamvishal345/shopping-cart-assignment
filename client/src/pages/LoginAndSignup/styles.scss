@import "assets/scss/_breakpoints.scss";
.login-signup-container {
  display: flex;
  padding: 2rem 0rem;
  justify-content: space-between;
  margin: 0 10rem;
  @include lg {
    margin: 0 4rem;
  }
  @include md {
    margin: 0 2rem;
  }
  @include md {
    margin: 0;
    flex-direction: column;
    align-items: center;
  }
  .details-section {
    h2 {
      margin-top: 0;
    }
    p {
      font-size: 12px;
      font-weight: 500;
    }
  }
  .form-section {
    width: 20rem;
    @include xm {
      width: 100%;
    }
    button {
      margin-top: 16px;
    }
    .error {
      font-size: 12px;
      margin-top: -10px;
      display: block;
      color: #f50606;
    }
  }
}
