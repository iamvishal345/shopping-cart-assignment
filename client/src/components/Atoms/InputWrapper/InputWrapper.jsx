import React from "react";
import "./style.scss";

export const InputWrapper = ({ value, setValue, name, id, label, type }) => (
  <div className="form-input">
    <input
      type={type || "text"}
      name={name}
      id={id}
      value={value}
      onChange={(e) => setValue(e.target.value.trimStart())}
    />
    <label
      className={`form-input-label ${value && "input-value"}`}
      htmlFor={id}
    >
      {label}
    </label>
  </div>
);

// InputField.propTypes = {
//   value: PropTypes.string.isRequired,
//   setValue: PropTypes.func.isRequired,
//   name: PropTypes.string.isRequired,
//   id: PropTypes.string.isRequired,
//   label: PropTypes.string.isRequired,
//   type: PropTypes.string,
//   ariaLabel: PropTypes.string,
// };
