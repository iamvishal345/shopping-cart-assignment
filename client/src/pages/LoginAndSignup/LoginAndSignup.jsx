import React, { useEffect, useState } from "react";
import "./styles.scss";
import { InputWrapper } from "components/Atoms/InputWrapper/InputWrapper";
import { Head } from "Utils/Head";
import { useHistory } from "react-router-dom";

const LoginAndSignup = ({ page }) => {
  const history = useHistory();
  const [form, setForm] = useState({});
  const [isValid, setIsValid] = useState({});
  useEffect(() => {
    setForm(
      page.fields.reduce((acc, field) => ({ ...acc, [field.name]: "" }), {})
    );
    setIsValid(
      page.fields.reduce((acc, field) => ({ ...acc, [field.name]: true }), {})
    );
  }, [page]);
  const handleValueChange = (value, name) => {
    setForm((form) => ({ ...form, [name]: value }));
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (validateForm()) {
      history.push("/");
    }
  };

  const validateForm = () => {
    let tempIsValid = { ...isValid };
    Object.keys(form).forEach((key) => {
      if (form[key] === "") {
        tempIsValid[key] = false;
      } else {
        tempIsValid[key] = validateValue(form[key], key);
      }
    });
    setIsValid(tempIsValid);
    return !Object.keys(tempIsValid).find((key) => !tempIsValid[key]);
  };

  const validateValue = (value, name) => {
    if (name === "email") {
      return /(.+)@(.+){2,}\.(.+){2,}/.test(value);
    }
    if (name === "confirmPassword") {
      return value === form["password"];
    }
    return true;
  };

  return (
    <React.Fragment>
      <Head
        title={`Sabka Bazaar - ${page.header}`}
        description={page.subHeader}
      />
      <div className="login-signup-container">
        <div className="details-section">
          <h2>{page.header}</h2>
          <p>{page.subHeader}</p>
        </div>
        <form className="form-section" onSubmit={handleSubmit}>
          {page.fields.map((field) => (
            <React.Fragment key={field.id}>
              <InputWrapper
                setValue={(value) => {
                  handleValueChange(value, field.name);
                }}
                value={form[field.name] || ""}
                {...field}
              />
              {!isValid[field.name] && (
                <small className="error">Invalid Value</small>
              )}
            </React.Fragment>
          ))}
          <button className="btn" type="submit">
            {page.header}
          </button>
        </form>
      </div>
    </React.Fragment>
  );
};

export default LoginAndSignup;
