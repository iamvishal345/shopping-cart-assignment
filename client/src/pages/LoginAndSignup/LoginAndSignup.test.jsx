import React from "react";
import { mount, shallow } from "enzyme";
import LoginAndSignup from "./LoginAndSignup";
import { REGISTER, SIGNIN } from "constants/constants";
import { InputWrapper } from "components/Atoms/InputWrapper/InputWrapper";

const mockHistory = jest.fn();
jest.mock("react-router-dom", () => {
  const ActualReact = jest.requireActual("react-router-dom");
  return {
    ...ActualReact,
    useHistory: () => ({ push: mockHistory }),
  };
});

describe("Login and Signup", () => {
  it("should match the snapshot with login page", () => {
    const wrapped = mount(<LoginAndSignup page={REGISTER} />);
    expect(wrapped).toMatchSnapshot();
  });
  it("should match the snapshot with signIn Page", () => {
    const wrapped = mount(<LoginAndSignup page={SIGNIN} />);
    expect(wrapped).toMatchSnapshot();
  });

  it("should handle InputWrapper value change callback", () => {
    const wrapped = shallow(<LoginAndSignup page={SIGNIN} />);
    wrapped.find(InputWrapper).at(0).prop("setValue")("testValue");
    expect(wrapped.find(InputWrapper).at(0).prop("value")).toEqual("testValue");
  });

  it("should submit Invalid form and show error messages", () => {
    const wrapped = shallow(<LoginAndSignup page={SIGNIN} />);
    wrapped.find(InputWrapper).at(0).prop("setValue")("");
    wrapped.find(InputWrapper).at(1).prop("setValue")("");
    wrapped.find("form").simulate("submit", { preventDefault: jest.fn() });
    expect(wrapped.find("small").at(0).text()).toEqual("Invalid Value");
    expect(wrapped.find("small").at(1).text()).toEqual("Invalid Value");
  });

  it("should submit valid form and redirect to home page", () => {
    const wrapped = shallow(<LoginAndSignup page={SIGNIN} />);
    wrapped.find(InputWrapper).at(0).prop("setValue")("abc@example.com");
    wrapped.find(InputWrapper).at(1).prop("setValue")("1234@1234");
    wrapped.find("form").simulate("submit", { preventDefault: jest.fn() });
    expect(mockHistory).toBeCalledWith("/");
  });
});
