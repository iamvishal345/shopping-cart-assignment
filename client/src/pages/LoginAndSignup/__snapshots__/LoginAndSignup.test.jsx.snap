// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login and Signup should match the snapshot with login page 1`] = `
<LoginAndSignup
  page={
    Object {
      "fields": Array [
        Object {
          "id": "firstName",
          "label": "FirstName",
          "name": "firstName",
          "type": "text",
        },
        Object {
          "id": "lastName",
          "label": "LastName",
          "name": "lastName",
          "type": "text",
        },
        Object {
          "id": "email",
          "label": "Email",
          "name": "email",
          "type": "email",
        },
        Object {
          "id": "password",
          "label": "Password",
          "name": "password",
          "type": "password",
        },
        Object {
          "id": "confirmPassword",
          "label": "Confirm Password",
          "name": "confirmPassword",
          "type": "password",
        },
      ],
      "header": "Signup",
      "subHeader": "We do not share your personal details with anyone",
    }
  }
>
  <Head
    description="We do not share your personal details with anyone"
    title="Sabka Bazaar - Signup"
  >
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <SideEffect(NullComponent)
        defer={true}
        encodeSpecialCharacters={true}
        meta={
          Array [
            Object {
              "content": "Sabka Bazaar - Signup",
              "name": "title",
            },
            Object {
              "content": "We do not share your personal details with anyone",
              "name": "description",
            },
            Object {
              "content": "website",
              "property": "og:type",
            },
            Object {
              "content": "en_US",
              "property": "og:locale",
            },
            Object {
              "content": "http://sabkabazaar.com/",
              "property": "og:url",
            },
            Object {
              "content": "Sabka Bazaar - Signup",
              "property": "og:title",
            },
            Object {
              "content": "We do not share your personal details with anyone",
              "property": "og:description",
            },
            Object {
              "content": "/static/images/logo_2x.png",
              "property": "og:image",
            },
            Object {
              "content": "Sabka Bazaar - Buy fresh Fruits, Vegetables, groceries and beauty products.",
              "property": "twitter:card",
            },
            Object {
              "content": "http://sabkabazaar.com/",
              "property": "twitter:url",
            },
            Object {
              "content": "Sabka Bazaar - Buy fresh Fruits, Vegetables, groceries and beauty products.",
              "property": "twitter:title",
            },
            Object {
              "content": "We do not share your personal details with anyone",
              "property": "twitter:description",
            },
            Object {
              "content": "/static/images/logo_2x.png",
              "property": "twitter:image",
            },
          ]
        }
        title="Sabka Bazaar - Signup"
        titleAttributes={Object {}}
      >
        <NullComponent
          defer={true}
          encodeSpecialCharacters={true}
          meta={
            Array [
              Object {
                "content": "Sabka Bazaar - Signup",
                "name": "title",
              },
              Object {
                "content": "We do not share your personal details with anyone",
                "name": "description",
              },
              Object {
                "content": "website",
                "property": "og:type",
              },
              Object {
                "content": "en_US",
                "property": "og:locale",
              },
              Object {
                "content": "http://sabkabazaar.com/",
                "property": "og:url",
              },
              Object {
                "content": "Sabka Bazaar - Signup",
                "property": "og:title",
              },
              Object {
                "content": "We do not share your personal details with anyone",
                "property": "og:description",
              },
              Object {
                "content": "/static/images/logo_2x.png",
                "property": "og:image",
              },
              Object {
                "content": "Sabka Bazaar - Buy fresh Fruits, Vegetables, groceries and beauty products.",
                "property": "twitter:card",
              },
              Object {
                "content": "http://sabkabazaar.com/",
                "property": "twitter:url",
              },
              Object {
                "content": "Sabka Bazaar - Buy fresh Fruits, Vegetables, groceries and beauty products.",
                "property": "twitter:title",
              },
              Object {
                "content": "We do not share your personal details with anyone",
                "property": "twitter:description",
              },
              Object {
                "content": "/static/images/logo_2x.png",
                "property": "twitter:image",
              },
            ]
          }
          title="Sabka Bazaar - Signup"
          titleAttributes={Object {}}
        />
      </SideEffect(NullComponent)>
    </HelmetWrapper>
  </Head>
  <div
    className="login-signup-container"
  >
    <div
      className="details-section"
    >
      <h2>
        Signup
      </h2>
      <p>
        We do not share your personal details with anyone
      </p>
    </div>
    <form
      className="form-section"
      onSubmit={[Function]}
    >
      <InputWrapper
        id="firstName"
        label="FirstName"
        name="firstName"
        setValue={[Function]}
        type="text"
        value=""
      >
        <div
          className="form-input"
        >
          <input
            id="firstName"
            name="firstName"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label
            className="form-input-label "
            htmlFor="firstName"
          >
            FirstName
          </label>
        </div>
      </InputWrapper>
      <InputWrapper
        id="lastName"
        label="LastName"
        name="lastName"
        setValue={[Function]}
        type="text"
        value=""
      >
        <div
          className="form-input"
        >
          <input
            id="lastName"
            name="lastName"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label
            className="form-input-label "
            htmlFor="lastName"
          >
            LastName
          </label>
        </div>
      </InputWrapper>
      <InputWrapper
        id="email"
        label="Email"
        name="email"
        setValue={[Function]}
        type="email"
        value=""
      >
        <div
          className="form-input"
        >
          <input
            id="email"
            name="email"
            onChange={[Function]}
            type="email"
            value=""
          />
          <label
            className="form-input-label "
            htmlFor="email"
          >
            Email
          </label>
        </div>
      </InputWrapper>
      <InputWrapper
        id="password"
        label="Password"
        name="password"
        setValue={[Function]}
        type="password"
        value=""
      >
        <div
          className="form-input"
        >
          <input
            id="password"
            name="password"
            onChange={[Function]}
            type="password"
            value=""
          />
          <label
            className="form-input-label "
            htmlFor="password"
          >
            Password
          </label>
        </div>
      </InputWrapper>
      <InputWrapper
        id="confirmPassword"
        label="Confirm Password"
        name="confirmPassword"
        setValue={[Function]}
        type="password"
        value=""
      >
        <div
          className="form-input"
        >
          <input
            id="confirmPassword"
            name="confirmPassword"
            onChange={[Function]}
            type="password"
            value=""
          />
          <label
            className="form-input-label "
            htmlFor="confirmPassword"
          >
            Confirm Password
          </label>
        </div>
      </InputWrapper>
      <button
        className="btn"
        type="submit"
      >
        Signup
      </button>
    </form>
  </div>
</LoginAndSignup>
`;

exports[`Login and Signup should match the snapshot with signIn Page 1`] = `
<LoginAndSignup
  page={
    Object {
      "fields": Array [
        Object {
          "id": "email",
          "label": "Email",
          "name": "email",
          "type": "email",
        },
        Object {
          "id": "password",
          "label": "Password",
          "name": "password",
          "type": "password",
        },
      ],
      "header": "Login",
      "subHeader": "Get access to your Orders, Wishlist and Recommendayions",
    }
  }
>
  <Head
    description="Get access to your Orders, Wishlist and Recommendayions"
    title="Sabka Bazaar - Login"
  >
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <SideEffect(NullComponent)
        defer={true}
        encodeSpecialCharacters={true}
        meta={
          Array [
            Object {
              "content": "Sabka Bazaar - Login",
              "name": "title",
            },
            Object {
              "content": "Get access to your Orders, Wishlist and Recommendayions",
              "name": "description",
            },
            Object {
              "content": "website",
              "property": "og:type",
            },
            Object {
              "content": "en_US",
              "property": "og:locale",
            },
            Object {
              "content": "http://sabkabazaar.com/",
              "property": "og:url",
            },
            Object {
              "content": "Sabka Bazaar - Login",
              "property": "og:title",
            },
            Object {
              "content": "Get access to your Orders, Wishlist and Recommendayions",
              "property": "og:description",
            },
            Object {
              "content": "/static/images/logo_2x.png",
              "property": "og:image",
            },
            Object {
              "content": "Sabka Bazaar - Buy fresh Fruits, Vegetables, groceries and beauty products.",
              "property": "twitter:card",
            },
            Object {
              "content": "http://sabkabazaar.com/",
              "property": "twitter:url",
            },
            Object {
              "content": "Sabka Bazaar - Buy fresh Fruits, Vegetables, groceries and beauty products.",
              "property": "twitter:title",
            },
            Object {
              "content": "Get access to your Orders, Wishlist and Recommendayions",
              "property": "twitter:description",
            },
            Object {
              "content": "/static/images/logo_2x.png",
              "property": "twitter:image",
            },
          ]
        }
        title="Sabka Bazaar - Login"
        titleAttributes={Object {}}
      >
        <NullComponent
          defer={true}
          encodeSpecialCharacters={true}
          meta={
            Array [
              Object {
                "content": "Sabka Bazaar - Login",
                "name": "title",
              },
              Object {
                "content": "Get access to your Orders, Wishlist and Recommendayions",
                "name": "description",
              },
              Object {
                "content": "website",
                "property": "og:type",
              },
              Object {
                "content": "en_US",
                "property": "og:locale",
              },
              Object {
                "content": "http://sabkabazaar.com/",
                "property": "og:url",
              },
              Object {
                "content": "Sabka Bazaar - Login",
                "property": "og:title",
              },
              Object {
                "content": "Get access to your Orders, Wishlist and Recommendayions",
                "property": "og:description",
              },
              Object {
                "content": "/static/images/logo_2x.png",
                "property": "og:image",
              },
              Object {
                "content": "Sabka Bazaar - Buy fresh Fruits, Vegetables, groceries and beauty products.",
                "property": "twitter:card",
              },
              Object {
                "content": "http://sabkabazaar.com/",
                "property": "twitter:url",
              },
              Object {
                "content": "Sabka Bazaar - Buy fresh Fruits, Vegetables, groceries and beauty products.",
                "property": "twitter:title",
              },
              Object {
                "content": "Get access to your Orders, Wishlist and Recommendayions",
                "property": "twitter:description",
              },
              Object {
                "content": "/static/images/logo_2x.png",
                "property": "twitter:image",
              },
            ]
          }
          title="Sabka Bazaar - Login"
          titleAttributes={Object {}}
        />
      </SideEffect(NullComponent)>
    </HelmetWrapper>
  </Head>
  <div
    className="login-signup-container"
  >
    <div
      className="details-section"
    >
      <h2>
        Login
      </h2>
      <p>
        Get access to your Orders, Wishlist and Recommendayions
      </p>
    </div>
    <form
      className="form-section"
      onSubmit={[Function]}
    >
      <InputWrapper
        id="email"
        label="Email"
        name="email"
        setValue={[Function]}
        type="email"
        value=""
      >
        <div
          className="form-input"
        >
          <input
            id="email"
            name="email"
            onChange={[Function]}
            type="email"
            value=""
          />
          <label
            className="form-input-label "
            htmlFor="email"
          >
            Email
          </label>
        </div>
      </InputWrapper>
      <InputWrapper
        id="password"
        label="Password"
        name="password"
        setValue={[Function]}
        type="password"
        value=""
      >
        <div
          className="form-input"
        >
          <input
            id="password"
            name="password"
            onChange={[Function]}
            type="password"
            value=""
          />
          <label
            className="form-input-label "
            htmlFor="password"
          >
            Password
          </label>
        </div>
      </InputWrapper>
      <button
        className="btn"
        type="submit"
      >
        Login
      </button>
    </form>
  </div>
</LoginAndSignup>
`;
